File: python/lombokalgoritma/compression/lz77.py
# LombokAlgoritma — LZ77, byte-oriented, window <= 255 (SPEC §13.3)
# SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
from __future__ import annotations

from ..errors import InvalidInputError, OutOfRangeError


def lz77_compress(data: bytes | bytearray, window: int = 255) -> bytes:
    """Tokens ``00 lit`` / ``01 off len`` (``3 <= len <= 255``, ``1 <= off <= window``).

    Greedy longest match; among equally long matches the largest offset wins.

    Raises:
        OutOfRangeError: ``window`` not an integer in [1, 255].
    """
    if isinstance(window, bool) or not isinstance(window, int) or not 1 <= window <= 255:
        raise OutOfRangeError("lz77_compress: window must be an integer in [1, 255]")
    out = bytearray()
    i, n = 0, len(data)
    while i < n:
        best_len = best_off = 0
        for j in range(max(0, i - window), i):
            length = 0
            while i + length < n and data[j + length] == data[i + length] and length < 255:
                length += 1
            if length > best_len:
                best_len, best_off = length, i - j
        if best_len >= 3:
            out += bytes((1, best_off, best_len))
            i += best_len
        else:
            out += bytes((0, data[i]))
            i += 1
    return bytes(out)


def lz77_decompress(data: bytes | bytearray) -> bytes:
    """Inverse of :func:`lz77_compress`.

    Raises:
        InvalidInputError: bad flag, truncated token, or an offset of 0 / beyond the output.
    """
    out = bytearray()
    i, n = 0, len(data)
    while i < n:
        flag = data[i]
        i += 1
        if flag == 0:
            if i >= n:
                raise InvalidInputError("lz77_decompress: truncated literal")
            out.append(data[i])
            i += 1
        elif flag == 1:
            if i + 1 >= n:
                raise InvalidInputError("lz77_decompress: truncated match")
            off, length = data[i], data[i + 1]
            i += 2
            if off == 0 or off > len(out):
                raise InvalidInputError("lz77_decompress: bad offset")
            start = len(out) - off
            for j in range(length):
                out.append(out[start + j])
        else:
            raise InvalidInputError(f"lz77_decompress: bad token flag {flag}")
    return bytes(out)

File: rust/lombokalgoritma/src/compression/lz77.rs
// LombokAlgoritma — LZ77 (SPEC §13.3)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
use crate::{Error, Result};
use alloc::vec::Vec;

/// Default (and maximum) window size.
pub const LZ77_MAX_WINDOW: usize = 255;

/// LZ77 with tokens `00 literal` or `01 offset length` (`1 ≤ offset ≤ window ≤ 255`,
/// `3 ≤ length ≤ 255`, matches may overlap). Greedy longest match; among equally long matches
/// the earliest start (largest offset) wins.
///
/// # Errors
/// [`Error::OutOfRange`] when `window ∉ [1, 255]`.
pub fn lz77_compress(data: &[u8], window: usize) -> Result<Vec<u8>> {
    if window == 0 || window > LZ77_MAX_WINDOW {
        return Err(Error::OutOfRange);
    }
    let mut out = Vec::new();
    let mut i = 0;
    while i < data.len() {
        let (mut best_len, mut best_off) = (0usize, 0usize);
        for j in i.saturating_sub(window)..i {
            let mut len = 0;
            while i + len < data.len() && data[j + len] == data[i + len] && len < 255 {
                len += 1;
            }
            if len > best_len {
                best_len = len;
                best_off = i - j;
            }
        }
        if best_len >= 3 {
            out.extend_from_slice(&[1, best_off as u8, best_len as u8]);
            i += best_len;
        } else {
            out.extend_from_slice(&[0, data[i]]);
            i += 1;
        }
    }
    Ok(out)
}

/// Inverse of [`lz77_compress`].
///
/// # Errors
/// [`Error::InvalidInput`] for a flag ∉ {0, 1}, a truncated token, or an offset of 0 or beyond
/// the output produced so far.
pub fn lz77_decompress(data: &[u8]) -> Result<Vec<u8>> {
    let mut out: Vec<u8> = Vec::new();
    let mut i = 0;
    while i < data.len() {
        let flag = data[i];
        i += 1;
        match flag {
            0 => {
                let &lit = data.get(i).ok_or(Error::InvalidInput)?;
                out.push(lit);
                i += 1;
            }
            1 => {
                if i + 1 >= data.len() {
                    return Err(Error::InvalidInput);
                }
                let (off, len) = (usize::from(data[i]), usize::from(data[i + 1]));
                i += 2;
                if off == 0 || off > out.len() {
                    return Err(Error::InvalidInput);
                }
                let start = out.len() - off;
                for j in 0..len {
                    out.push(out[start + j]);
                }
            }
            _ => return Err(Error::InvalidInput),
        }
    }
    Ok(out)
}
